File: cocos2dx-js/Poco/sdk/AbstractNode.js
var AbstractNode = function () {
}

AbstractNode.prototype.getParent = function () {
    return null
}
    
AbstractNode.prototype.getChildren = function () {
    // :rettype: Iterable<AbstractNode>
}
    
AbstractNode.prototype.getAttr = function (attrName) {
    var attrs = {
        name: '<Root>',
        type: 'Root',
        visible: true,
        pos: [0.0, 0.0],
        size: [0.0, 0.0],
        scale: [1.0, 1.0],
        anchorPoint: [0.5, 0.5],
        zOrders: {'local': 0, 'global': 0},
    }

    return attrs[attrName]
}
    
AbstractNode.prototype.setAttr = function (attrName, val) {
    throw new Error('Unable to set attributes ' + attrName + ' on this node. (NotImplemented) ')
}

AbstractNode.prototype.getAvailableAttributeNames = function () {
    // enumerate all available attributes' name of this node
    // :rettype: Iterable<string>

    return [
        "name",
        "type",
        "visible",
        "pos",
        "size",
        "scale",
        "anchorPoint",
        "zOrders",
    ]
}

AbstractNode.prototype.enumerateAttrs = function () {
    // :rettype: Iterable<string, ValueType>
    var ret = {}
    var allAttrNames = this.getAvailableAttributeNames()
    for (var i in allAttrNames) {
        var attrName = allAttrNames[i]
        var attrVal = this.getAttr(attrName)
        if (attrVal !== undefined) {
            ret[attrName] = attrVal
        }
    }
    return ret
}

try {
    module.exports = AbstractNode;
} catch (e) {
    if (window.module && window.module.exports) {
        window.module.exports = AbstractNode;
    }
}

File: cocos2dx-lua/poco/sdk/AbstractNode.lua
local AbstractNode = {}
AbstractNode.__index = AbstractNode
AbstractNode.__isPocoNodeWrapper__ = true

function AbstractNode:getParent()
    --
    -- :rettype: AbstractNode or nil
    
    return nil
end
    
function AbstractNode:getChildren()
    -- 
    -- :rettype: Iterable<AbstractNode>
end

function AbstractNode:getAvailableAttributeNames()
    -- get available attributes of this node
    -- :rettype: Iterable<string>

    return {
        "name",
        "type",
        "visible",
        "pos",
        "size",
        "scale",
        "anchorPoint",
        "zOrders",
    }
end

function AbstractNode:getAttr(attrName)
    -- 
    -- :rettype: <any>

    local attrs = {
        name = '<Root>',
        type = 'Root',
        visible = true,
        pos = {0.0, 0.0},
        size = {0.0, 0.0},
        scale = {1.0, 1.0},
        anchorPoint = {0.5, 0.5},
        zOrders = {['local'] = 0, ['global'] = 0},
    }
    return attrs[attrName]
end

function AbstractNode:setAttr(attrName, val)
    --
    -- :retval: true if success else false
    
    assert(false, string.format('unable to set attributes "%s" on this node', attrName))
    return false
end
    
function AbstractNode:enumerateAttrs()
    --
    -- :rettype: Iterable<string, ValueType>

    local attrs = {}
    for _, attrName in ipairs(self:getAvailableAttributeNames()) do
        local attrVal = self:getAttr(attrName)
        if attrVal ~= nil then
            attrs[attrName] = attrVal
        end
    end
    return attrs
end

return AbstractNode